### ISSUE TITLE
Iterate over multiple partition key table with Cassandra DataStax Driver

### ISSUE THREAD

**Issue by vanessaschissato** [Impetus/kundera#614]
I have this table with a multiple partition key:

``` sql
CREATE TABLE replication (
    core text,
    date_hour text,
    content text
    PRIMARY KEY ((core, date_hour), content)
);
```

And i'm trying iterate like this:

``` java
com.impetus.kundera.query.Query query = (com.impetus.kundera.query.Query) this.entityManagerProvider
                .get()
                .createQuery(
                        "SELECT entityName FROM " + this.getClass() + " entityName ");

query.setFetchSize(3);

return query.iterate();
```

But, due to my table having composite partition keys, i'm got this error:

``` java
Caused by: com.impetus.kundera.KunderaException: com.datastax.driver.core.exceptions.InvalidQueryException: Invalid number of arguments in call to function token: 2 required but 1 provided
    at com.impetus.kundera.client.cassandra.dsdriver.DSClient.execute(DSClient.java:525)
    at com.impetus.kundera.client.cassandra.dsdriver.DSClient.executeQuery(DSClient.java:432)
    at com.impetus.client.cassandra.query.ResultIterator.populateEntities(ResultIterator.java:251)
    at com.impetus.client.cassandra.query.ResultIterator.onCheckRelation(ResultIterator.java:195)
    ... 24 more
Caused by: com.datastax.driver.core.exceptions.InvalidQueryException: Invalid number of arguments in call to function token: 2 required but 1 provided
    at com.datastax.driver.core.exceptions.InvalidQueryException.copy(InvalidQueryException.java:35)
    at com.datastax.driver.core.DefaultResultSetFuture.extractCauseFromExecutionException(DefaultResultSetFuture.java:256)
    at com.datastax.driver.core.DefaultResultSetFuture.getUninterruptibly(DefaultResultSetFuture.java:172)
    at com.datastax.driver.core.SessionManager.execute(SessionManager.java:80)
    at com.impetus.kundera.client.cassandra.dsdriver.DSClient.execute(DSClient.java:520)
    ... 27 more
Caused by: com.datastax.driver.core.exceptions.InvalidQueryException: Invalid number of arguments in call to function token: 2 required but 1 provided
    at com.datastax.driver.core.Responses$Error.asException(Responses.java:96)
    at com.datastax.driver.core.DefaultResultSetFuture.onSet(DefaultResultSetFuture.java:108)
    at com.datastax.driver.core.RequestHandler.setFinalResult(RequestHandler.java:228)
    at com.datastax.driver.core.RequestHandler.onSet(RequestHandler.java:354)
    at com.datastax.driver.core.Connection$Dispatcher.messageReceived(Connection.java:561)
    at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
    at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
    at org.jboss.netty.handler.codec.oneone.OneToOneDecoder.handleUpstream(OneToOneDecoder.java:70)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
    at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
    at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
    at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462)
    at org.jboss.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443)
    at org.jboss.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:303)
    at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
    at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
    at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
    at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
    at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
    at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:107)
    at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:312)
    at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:88)
    at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
    at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
    at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
    ... 3 more
```

**Comment by chhavigangwal** [comment]
Can you share the sample entity corresponding to the table created ?

Chhavi

**Comment by vanessaschissato** [comment]
``` java
@Entity
@Table(name = "replication")
public class Replication extends BaseEntity<ReplicationId> {

    private static final long serialVersionUID = -8055485531216607863L;

    @EmbeddedId
         private ReplicationId id;

    // --

    public Replication() {
        id = new ReplicationId();
    }

    // --

    public ReplicationId getId() {
        return id;
    }

    public void setId(ReplicationId id) {
        this.id = id;
    }
}
```

``` java
@Embeddable
public class ReplicationId implements Serializable
{

    @Column(name = "core")
    private String core;

    @Column(name = "date_hour")
    private String dateHour;

    @Column(name = "content")
    private String content;

    public String getCore() {
        return core;
    }

    public void setCore(String core) {
        this.core = core;
    }

    public String getDateHour() {
        return dateHour;
    }

    public void setDateHour(String dateHour) {
        this.dateHour = dateHour;
    }

    public String getContent() {
        return content;
    }

    public void setContent(String content) {
        this.content = content;
    }
}
```

**Comment by chhavigangwal** [comment]
Two things here :  
 1)  I tried the same with Cassandra 2.0.4 and Kundera's trunk code and it is iterating over the results without any exception. What is the content of your BaseEntity ? This will help me debugging issue further.
 2)  For multiple partition key you need to create EmbeddedId composite key field 
Sample entity definition can be found at : 
https://github.com/impetus-opensource/Kundera/blob/d9d8772c9e138cfcd349aa08050a1000c2532469/src/kundera-cassandra/cassandra-core/src/test/java/com/impetus/client/crud/compositeType/EntityWithMultiplePartitionKey.java
Corresponding test case :
https://github.com/impetus-opensource/Kundera/blob/a2e0122776f1a166d8aadf80cbbe418500bd4605/src/kundera-cassandra/cassandra-core/src/test/java/com/impetus/client/crud/compositeType/EntityWithMultiplePartitionKeyTest.java

Chhavi

**Comment by vanessaschissato** [comment]
Thank you.

**Comment by vanessaschissato** [comment]
I mapped my entities like suggested above, but now i'm getting a Kundera Exception.
I'm using Kundera 2.12.

``` java
@Entity
@Table(name = "replication")
public class Replication extends BaseEntity<ReplicationId> {

    private static final long serialVersionUID = -8055485531216607863L;

    @EmbeddedId
         private ReplicationId id;

    // --

    public Replication() {
        id = new ReplicationId();
    }

    // --

    public ReplicationId getId() {
        return id;
    }

    public void setId(ReplicationId id) {
        this.id = id;
    }
}
```

``` java
@Embeddable
public class ReplicationId implements Serializable
{

     @EmbeddedId
      private ReplicationPartitionKey partitionKey;

      public ReplicationId() {
          id = new ReplicationPartitionKey();
      }

      @Column(name = "content")
      private String content;

       public SolrReplicationPartitionKey getPartitionKey() {
        return partitionKey;
    }

    public void setPartitionKey(SolrReplicationPartitionKey partitionKey) {
        this.partitionKey = partitionKey;
    }

    public String getContent() {
        return content;
    }

    public void setContent(String content) {
        this.content = content;
    }
}
```

``` java
@Embeddable
public class ReplicationPartitionKey implements Serializable
{

    @Column(name = "core")
    private String core;

    @Column(name = "date_hour")
    private String dateHour;

     public String getCore() {
        return core;
    }

    public void setCore(String core) {
        this.core = core;
    }

    public String getDateHour() {
        return dateHour;
    }

    public void setDateHour(String dateHour) {
        this.dateHour = dateHour;
    }
}
```

``` java
Caused by: java.lang.ClassCastException: com.impetus.kundera.metadata.model.type.DefaultEmbeddableType cannot be cast to com.impetus.kundera.metadata.model.type.AbstractIdentifiableType
    at com.impetus.kundera.metadata.processor.MetaModelBuilder$TypeBuilder$AttributeBuilder.build(MetaModelBuilder.java:463)
    at com.impetus.kundera.metadata.processor.MetaModelBuilder$TypeBuilder.build(MetaModelBuilder.java:358)
    at com.impetus.kundera.metadata.processor.MetaModelBuilder.onDeclaredFields(MetaModelBuilder.java:801)
    at com.impetus.kundera.metadata.processor.MetaModelBuilder.buildManagedType(MetaModelBuilder.java:700)
    at com.impetus.kundera.metadata.processor.MetaModelBuilder.processInternal(MetaModelBuilder.java:817)
    at com.impetus.kundera.metadata.processor.MetaModelBuilder.getType(MetaModelBuilder.java:758)
    at com.impetus.kundera.metadata.processor.MetaModelBuilder.access$400(MetaModelBuilder.java:80)
    at com.impetus.kundera.metadata.processor.MetaModelBuilder$TypeBuilder$AttributeBuilder.onSuperType(MetaModelBuilder.java:495)
    at com.impetus.kundera.metadata.processor.MetaModelBuilder$TypeBuilder$AttributeBuilder.build(MetaModelBuilder.java:461)
    at com.impetus.kundera.metadata.processor.MetaModelBuilder$TypeBuilder.build(MetaModelBuilder.java:358)
    at com.impetus.kundera.metadata.processor.MetaModelBuilder.construct(MetaModelBuilder.java:140)
    at com.impetus.kundera.metadata.processor.TableProcessor.populateMetadata(TableProcessor.java:134)
    at com.impetus.kundera.metadata.processor.TableProcessor.process(TableProcessor.java:99)
    at com.impetus.kundera.metadata.MetadataBuilder.buildEntityMetadata(MetadataBuilder.java:133)
    at com.impetus.kundera.configure.MetamodelConfiguration.scanClassAndPutMetadata(MetamodelConfiguration.java:365)
    at com.impetus.kundera.configure.MetamodelConfiguration.loadEntityMetadata(MetamodelConfiguration.java:251)
    at com.impetus.kundera.configure.MetamodelConfiguration.configure(MetamodelConfiguration.java:112)
    at com.impetus.kundera.persistence.EntityManagerFactoryImpl.configure(EntityManagerFactoryImpl.java:162)
    at com.impetus.kundera.persistence.EntityManagerFactoryImpl.<init>(EntityManagerFactoryImpl.java:139)
    at com.impetus.kundera.KunderaPersistence.createEntityManagerFactory(KunderaPersistence.java:85)
    at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:78)
```

**Comment by kkmishra** [comment]
Fix has been added into current development branch (trunk).
You can find related test case here
https://github.com/impetus-opensource/Kundera/blob/trunk/src/kundera-cassandra/cassandra-ds-driver/src/test/java/com/impetus/kundera/client/cassandra/composite/DSEntityWithMultiplePartitionKeyTest.java

Please verify.

Thanks
Kuldeep

**Comment by chhavigangwal** [comment]
Fixed releasing with 2.13
